File: ios/HybridViewModelInstance.swift
import RiveRuntime

class HybridViewModelInstance: HybridViewModelInstanceSpec {
  let viewModelInstance: RiveDataBindingViewModel.Instance?
  
  init(viewModelInstance: RiveDataBindingViewModel.Instance) {
    self.viewModelInstance = viewModelInstance
  }

  var instanceName: String { viewModelInstance?.name ?? "" }
  
  func numberProperty(path: String) throws -> (any HybridViewModelNumberPropertySpec)? {
    guard let property = viewModelInstance?.numberProperty(fromPath: path) else { return nil }
    return HybridViewModelNumberProperty(property: property)
  }
  
  func stringProperty(path: String) throws -> (any HybridViewModelStringPropertySpec)? {
    guard let property = viewModelInstance?.stringProperty(fromPath: path) else { return nil }
    return HybridViewModelStringProperty(property: property)
  }
  
  func booleanProperty(path: String) throws -> (any HybridViewModelBooleanPropertySpec)? {
    guard let property = viewModelInstance?.booleanProperty(fromPath: path) else { return nil }
    return HybridViewModelBooleanProperty(property: property)
  }
  
  func colorProperty(path: String) throws -> (any HybridViewModelColorPropertySpec)? {
    guard let property = viewModelInstance?.colorProperty(fromPath: path) else { return nil }
    return HybridViewModelColorProperty(property: property)
  }
  
  func enumProperty(path: String) throws -> (any HybridViewModelEnumPropertySpec)? {
    guard let property = viewModelInstance?.enumProperty(fromPath: path) else { return nil }
    return HybridViewModelEnumProperty(property: property)
  }
  
  func triggerProperty(path: String) throws -> (any HybridViewModelTriggerPropertySpec)? {
    guard let property = viewModelInstance?.triggerProperty(fromPath: path) else { return nil }
    return HybridViewModelTriggerProperty(property: property)
  }

  func imageProperty(path: String) throws -> (any HybridViewModelImagePropertySpec)? {
    guard let property = viewModelInstance?.imageProperty(fromPath: path) else { return nil }
    return HybridViewModelImageProperty(property: property)
  }

  func listProperty(path: String) throws -> (any HybridViewModelListPropertySpec)? {
    guard let property = viewModelInstance?.listProperty(fromPath: path) else { return nil }
    return HybridViewModelListProperty(property: property)
  }
}

File: android/src/experimental/java/com/margelo/nitro/rive/HybridViewModelInstance.kt
package com.margelo.nitro.rive

import android.util.Log
import androidx.annotation.Keep
import app.rive.ViewModelInstance
import app.rive.ViewModelInstanceSource
import app.rive.core.CommandQueue
import com.facebook.proguard.annotations.DoNotStrip
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking

@Keep
@DoNotStrip
class HybridViewModelInstance(
  internal val viewModelInstance: ViewModelInstance,
  private val riveWorker: CommandQueue,
  private val parentFile: HybridRiveFile,
  private val viewModelName: String? = null,
  private val _instanceName: String? = null
) : HybridViewModelInstanceSpec() {
  companion object {
    private const val TAG = "HybridViewModelInstance"
  }

  private val propertyNames: Set<String> by lazy {
    val name = viewModelName ?: return@lazy emptySet()
    val file = parentFile.riveFile ?: return@lazy emptySet()
    try {
      runBlocking { file.getViewModelProperties(name) }.map { it.name }.toSet()
    } catch (e: Exception) {
      Log.e(TAG, "Failed to fetch property names for viewModel '$name'", e)
      emptySet()
    }
  }

  private fun hasProperty(path: String): Boolean {
    if (propertyNames.isEmpty()) return true
    return propertyNames.contains(path)
  }

  override val instanceName: String
    get() = _instanceName ?: ""

  override fun numberProperty(path: String): HybridViewModelNumberPropertySpec? {
    return try {
      runBlocking { viewModelInstance.getNumberFlow(path).first() }
      HybridViewModelNumberProperty(viewModelInstance, path)
    } catch (e: Exception) {
      Log.e(TAG, "numberProperty failed for path '$path'", e)
      null
    }
  }

  override fun stringProperty(path: String): HybridViewModelStringPropertySpec? {
    return try {
      runBlocking { viewModelInstance.getStringFlow(path).first() }
      HybridViewModelStringProperty(viewModelInstance, path)
    } catch (e: Exception) {
      Log.e(TAG, "stringProperty failed for path '$path'", e)
      null
    }
  }

  override fun booleanProperty(path: String): HybridViewModelBooleanPropertySpec? {
    return try {
      runBlocking { viewModelInstance.getBooleanFlow(path).first() }
      HybridViewModelBooleanProperty(viewModelInstance, path)
    } catch (e: Exception) {
      Log.e(TAG, "booleanProperty failed for path '$path'", e)
      null
    }
  }

  override fun colorProperty(path: String): HybridViewModelColorPropertySpec? {
    return try {
      runBlocking { viewModelInstance.getColorFlow(path).first() }
      HybridViewModelColorProperty(viewModelInstance, path)
    } catch (e: Exception) {
      Log.e(TAG, "colorProperty failed for path '$path'", e)
      null
    }
  }

  override fun enumProperty(path: String): HybridViewModelEnumPropertySpec? {
    return try {
      runBlocking { viewModelInstance.getEnumFlow(path).first() }
      HybridViewModelEnumProperty(viewModelInstance, path)
    } catch (e: Exception) {
      Log.e(TAG, "enumProperty failed for path '$path'", e)
      null
    }
  }

  override fun triggerProperty(path: String): HybridViewModelTriggerPropertySpec? {
    if (!hasProperty(path)) return null
    return try {
      HybridViewModelTriggerProperty(viewModelInstance, path)
    } catch (e: Exception) {
      Log.e(TAG, "triggerProperty failed for path '$path'", e)
      null
    }
  }

  override fun imageProperty(path: String): HybridViewModelImagePropertySpec? {
    return try {
      HybridViewModelImageProperty(viewModelInstance, path, riveWorker)
    } catch (e: Exception) {
      Log.e(TAG, "imageProperty failed for path '$path'", e)
      null
    }
  }

  override fun listProperty(path: String): HybridViewModelListPropertySpec? {
    return try {
      HybridViewModelListProperty(viewModelInstance, path, riveWorker, parentFile)
    } catch (e: Exception) {
      Log.e(TAG, "listProperty failed for path '$path'", e)
      null
    }
  }

  override fun artboardProperty(path: String): HybridViewModelArtboardPropertySpec? {
    return try {
      HybridViewModelArtboardProperty(viewModelInstance, path, parentFile)
    } catch (e: Exception) {
      Log.e(TAG, "artboardProperty failed for path '$path'", e)
      null
    }
  }

  override fun viewModel(path: String): HybridViewModelInstanceSpec? {
    if (!hasProperty(path)) return null
    return try {
      val file = parentFile.riveFile ?: return null
      val source = ViewModelInstanceSource.Reference(viewModelInstance, path)
      val childVmi = ViewModelInstance.fromFile(file, source)
      HybridViewModelInstance(childVmi, riveWorker, parentFile)
    } catch (e: Exception) {
      Log.e(TAG, "viewModel failed for path '$path'", e)
      null
    }
  }

  override fun replaceViewModel(path: String, instance: HybridViewModelInstanceSpec) {
    Log.w(TAG, "replaceViewModel not yet supported in experimental API")
  }
}
